stdlib/random.ml

```diff
@@ -114,35 +114,112 @@ module State = struct
+     * This is what the test below does (while carefuly avoiding
+     * overflows and sparing a division [max_int / n]). *)
+    if r - v > max_int - n + 1 then full_int_aux s n else v
 
   let full_int s bound =
     if bound <= 0 then
       invalid_arg "Random.full_int"
+    (* When the bound fits in 31 bits, we use [int31aux]
+     * so as to yield the same output on 32-bit and 64-bit platforms. *)
     else if bound > 0x3FFFFFFF then
-      int63aux s bound
+      full_int_aux s bound
+    else
+      int31aux s bound
+
+  (* Cast a full-width integer to a signed 31-bit integer, by computing its
+   * signed remainder modulo 2^31. *)
+  let int31_of_int x =
+    let d = Sys.int_size - 31 in
+    (* Replicate the 31th bit (sign bit) to higher positions: *)
+    (x lsl d) asr d
+
+  (* Return an integer between [min] (included) and [max] (included).
+     We must have [min <= max], and [min] and [max] must fit in 31 bits.
+     This function is designed so that it yields the same output on
+     32-bit and 64-bit platforms. *)
+  let rec int31_in_range_aux s ~min ~max =
+    let r = int31_of_int (Int64.to_int (next s)) in
+    if r < min || r > max then int31_in_range_aux s ~min ~max else r
+
+  let int31_in_range s ~min ~max =
+    let span = max - min + 1 in
+    (* Explanation of this test: see comment in [int_in_range];
+     * the comparison to 0 is for 32-bit platforms;
+     * the comparison to 2^30-1 is for 64-bit platforms. *)
+    if span <= 0 || span > 0x3FFF_FFFF then
+      int31_in_range_aux s ~min ~max
     else
-      intaux s bound
+      min + int31aux s span
+
+  (* Return an integer between [min] (included) and [max] (included).
+     We must have [min <= max]. *)
+  let rec int_in_range_aux s ~min ~max =
+    let r = Int64.to_int (next s) in
+    if r < min || r > max then int_in_range_aux s ~min ~max else r
+
+  let int_in_range s ~min ~max =
+    if min > max then
+      invalid_arg "Random.int_in_range"
+    (* When both bounds fit in 31 bits, we use [int31_in_range],
+     * so as to yield the same output on 32-bit and 64-bit platforms. *)
+    else if min < -0x4000_0000 || max > 0x3FFF_FFFF then
+      let span = max - min + 1 in
+      (* We use [int_in_range_aux] only if [max - min + 1] overflows;
+       * then, the interval between [min] and [max] covers at least half
+       * of the representable integers, so that it converges quickly
+       * (the probability of re-drawing is at most 1/2).
+       *
+       * When [max - min + 1] does not overflow, we use [full_int_aux],
+       * for 2 reasons:
+       *   - [int_in_range] then has a high probability of re-drawing,
+       *     so would be very slow;
+       *   - we thus guarantee that, when [min = 0], [int_in_range]
+       *     yields the same output as [full_int (max+1)].
+       *)
+      if span <= 0 then
+        int_in_range_aux s ~min ~max
+      else
+        min + full_int_aux s span
+    else
+      int31_in_range s ~min ~max
```

Comment:
I think this code would be clearer if either
- `int31_in_range` was inlined here, resulting in 4 cases
- a function `full_int_in_range`, similar to `int31_in_range` was introduced and called here, resulting in 2 cases.